core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala

```diff
@@ -166,17 +166,25 @@ private[spark] object ReliableCheckpointRDD extends Logging {
     val tempOutputPath =
       new Path(outputDir, s".$finalOutputName-attempt-${ctx.attemptNumber()}")
 
-    if (fs.exists(tempOutputPath)) {
-      throw new IOException(s"Checkpoint failed: temporary path $tempOutputPath already exists")
-    }
     val bufferSize = env.conf.getInt("spark.buffer.size", 65536)
 
-    val fileOutputStream = if (blockSize < 0) {
-      fs.create(tempOutputPath, false, bufferSize)
-    } else {
-      // This is mainly for testing purpose
-      fs.create(tempOutputPath, false, bufferSize,
-        fs.getDefaultReplication(fs.getWorkingDirectory), blockSize)
+    val fileOutputStream = try {
+      if (blockSize < 0) {
+        fs.create(tempOutputPath, false, bufferSize)
+      } else {
+        // This is mainly for testing purpose
+        fs.create(tempOutputPath, false, bufferSize,
+          fs.getDefaultReplication(fs.getWorkingDirectory), blockSize)
+      }
+    } catch {
+      case e: IOException =>
+        // could be FileAlreadyExistsException, or a more fundamental create failure
```

Comment:
changing to rely on the underlying layer to do things....create(overwrite=false) will trigger a FileAlreadyExistsException. Note that S3 and swift don't have an atomic check+create, but that's not avoidable.